### DISCUSSION TITLE
Subscribe to Single Node

### DISCUSSION THREAD

**Discussion by bibhudatta-sahu** [eclipse-milo/milo#886] [General]
Hello,

     I have multiple nodes in server. From those nodes, I would like to subscribe or unsubscribe to a single node (not all nodes) from client. I would like to know how can I achieve it through API. Please provide your suggestion. 
     
Thanks,
Bibhu

**Comment by arnoweiss** [comment]
Hi Bibhu,

[here's an example](https://github.com/eclipse/milo/blob/17a8fc8d14893e3e4532a9237789363b8cc50891/milo-examples/client-examples/src/main/java/org/eclipse/milo/examples/client/SubscriptionExample.java#L72) of how you can achieve your goal. You can basically just take the code, adjust the endpoint, port and NodeId and you're done.

**Reply by bibhudatta-sahu** [reply]
Thanks @alw-iwu . I will try it.

**Comment by kevinherron** [comment]
There is also a higher level ManagedSubscription and ManagedDataItem API: https://github.com/eclipse/milo/blob/17a8fc8d14893e3e4532a9237789363b8cc50891/milo-examples/client-examples/src/main/java/org/eclipse/milo/examples/client/ManagedSubscriptionDataExample.java

**Reply by bibhudatta-sahu** [reply]
Thanks @kevinherron . It is helpful.

**Comment by bibhudatta-sahu** [comment]
Hello Everyone,

Thank you for subscription. Now I am able to subscribe the nodes. But after performing the subscriptions, I would like to unsubscribe a particular node. 

For example:

I have subscribed to "node1" and "node2" and getting the subscription id "s1". Now I would like to unsubscribe only "node1" (not node2). I tried below approach to unsubscribe the node.

1. Storing node id and monitored item in a map during the time of subscription.
2. While doing unsubscribe, I am getting the monitored item id from map and calling below api to unsubscribe

**subscription.deleteMonitoredItems(item).get();**

But this approach didn't work for me. I am still able to see values are coming for the point which I unsubscribed.

Could you please help me how can I resolve the issue.

Thanks,
Bibhu

**Reply by kevinherron** [reply]
This should work in general. Without a code example it will be difficult to figure out where you've gone wrong.

**Reply by bibhudatta-sahu** [reply]
Hi @kevinherron ,

   I followed https://github.com/eclipse/milo/blob/2a9a23c562a0e23898df28e310a5cfae3870aabf/milo-examples/client-examples/src/main/java/org/eclipse/milo/examples/client/SubscriptionExample.java example to create subscription and monitored item. I created a map **Map<String, UaMonitoredItem> map = new ConcurrentHashMap<>();** to maintain the nodeid and monitored item. In the example, line number 91 (https://github.com/eclipse/milo/blob/2a9a23c562a0e23898df28e310a5cfae3870aabf/milo-examples/client-examples/src/main/java/org/eclipse/milo/examples/client/SubscriptionExample.java#L91), when item's status code is "Good", then I am storing nodeid and UaSubscription in map. Please refere below code 

map.put(item.getReadValueId().getNodeId().toParseableString(), item);

While doing the unsubscribe process. Code is as below:

UaMonitoredItem item = map.get(<NodeId>);
List<UaMonitoredItem> list=new ArrayList();
list.add(item);
StatusCode code = ubscription.deleteMonitoredItems(list).get();

and getting the status code good. But still I am getting the values for the node after doing unsubscribe.

**Reply by bibhudatta-sahu** [reply]
Yes, there was some wrong in the code. I was again creating Monitored Item for same node id in the unsubscribe flow. That's why there were two "Monitored Item" entries of same node got created for itemsByClientHandle object. So, though I was deleting the one monitored item object, still there was another monitored item object present in the itemsByClientHandle object. That's why I was again getting the values for the node (which I tried to unsubscribe). After removing the faulty code, now it is working as expected.
